### DISCUSSION TITLE
MDX not rendering properly when using Next.js with TypeScript and Tailwind CSS

### DISCUSSION THREAD

**Discussion by hbhutta** [mdx-js/mdx#2387] [Q&A] [answered]
Hi! I asked a very similar question on the `remarkjs` GitHub repository and got very detailed help from the repository owners. At that time I was having this same problem, but I was using the deprecated `npx create-react-app` utility alongside Tailwind and someone suggested that that was causing it and I should switch to a React framework with an up-to-date build utility.

So I'm using `Next.js` now, with Tailwind and TypeScript. I am trying to render some MDX pages. I am using the Pages router and my project structure looks like this:

![image](https://github.com/mdx-js/mdx/assets/112280419/e8b4a63d-f6e1-4326-8888-48637ddf9184)

I followed the official tutorial by Next.js [here](https://nextjs.org/docs/pages/building-your-application/configuring/mdx#remark-and-rehype-plugins), but the MDX is still not rendering properly (e.g. headers are not the expected size, no code blocks, etc..).

Here is my `next.config.js`:
```javascript
const withMDX = require('@next/mdx')()
 
/** @type {import('next').NextConfig} */
const nextConfig = {
  // Configure `pageExtensions`` to include MDX files
  pageExtensions: ['js', 'jsx', 'mdx', 'ts', 'tsx'],
  // Optionally, add any other Next.js config below
}
 
module.exports = withMDX(nextConfig)
```

My `tailwind.config.ts` also accommodates `.mdx` files:
```
import type { Config } from 'tailwindcss'

const config: Config = {
  content: [
    './pages/**/*.{js,ts,jsx,tsx,mdx}',
    './components/**/*.{js,ts,jsx,tsx,mdx}',
    './app/**/*.{js,ts,jsx,tsx,mdx}',
  ],
  plugins: [],
}
export default config;
```

Running `npm list` shows that I have all the necessary packages (all installed with zero vulnerabilities):

![image](https://github.com/mdx-js/mdx/assets/112280419/f8a5db14-6ed2-42b3-99cd-045173aee301)

I also checked the Next.js repository for a similar issue someone was having with MDX and Tailwind, but they resolved it because they only had a typo somewhere. I don't have any typos anywhere as far as I can tell. 

So what could be preventing the MDX files from rendering properly?

**Comment by ChristianMurphy** [comment]
Hey @hbhutta! 👋 
This was at least partially addressed in your last question https://github.com/orgs/remarkjs/discussions/1252#discussioncomment-7405197
Some solutions from other previous questions with solutions which may help you:
* https://github.com/remarkjs/react-markdown/issues/505#issuecomment-724911501
* https://github.com/orgs/remarkjs/discussions/681#discussioncomment-597356

**Reply by hbhutta** [reply]
Hi Christian, thank you for the links to previous questions. I've installed the typography package just in case. My question is more so about MDX and not about the `react-markdown` library. Specifically, my question is asked after I tried to implement the advice I got from you and the other maintainers in my previous question. 

I don't want to use `react-markdown` when I can use MDX in Next.js and when Next.js has great documentation for it.

The problem is that even following the tutorial/documentation, the MDX is still not rendering properly. I've even checked after configuring Tailwind typography.

**Reply by ChristianMurphy** [reply]
> the MDX is still not rendering properly (e.g. headers are not the expected size, no code blocks, etc..).

MDX does not provide any styles, if things are the wrong size, it is an issue with your stylesheets.

> My question is more so about MDX and not about the react-markdown library. 

Neither MDX nor react-markdown provide styles.
How to correctly configure tailwind so it renders markdown correctly applies to both.

> I don't want to use react-markdown when I can use MDX in Next.js and when Next.js has great documentation for it.

I'm not asking you to use react-markdown, I'm letting you know you need to configure tailwind, and sharing some configurations you could use.

> The problem is that even following the tutorial/documentation, the MDX is still not rendering properly. I've even checked after configuring Tailwind typography.

Again, see the above.
MDX does not provide styles, if the styles look wrong, triple check your stylesheet.

**Reply by hbhutta** [reply]
Hi Christian, thank you for your help so far. 

Here is my tailwind configuration at the moment:

```
import type { Config } from 'tailwindcss'

const config: Config = {
  content: [
    './pages/**/*.{js,ts,jsx,tsx,mdx}',
    './components/**/*.{js,ts,jsx,tsx,mdx}',
    './app/**/*.{js,ts,jsx,tsx,mdx}',
  ],
  plugins: [
    require('@tailwindcss/typography')
  ],
}
export default config;
```

And my stylesheet is just the `globals.css` that comes by default when using the `create-next-app` utility. It just has the three Tailwind directives, and is blank other than that: 
```css
@tailwind base;
@tailwind components;
@tailwind utilities;
```

Any styles I'm using on my components, I'm doing it through writing Tailwind utility classes directly in the `className` of that component, so I don't understand how Tailwind could be affecting the MDX rendering.

Do you have any clues as to what could be causing the problem?

**Reply by ChristianMurphy** [reply]
> I don't understand how Tailwind could be affecting the MDX rendering

Tailwind adds global styles to plain HTML elements.
MDX generates plain HTML elements, so is affected.

> Do you have any clues as to what could be causing the problem?

The global styles from tailwind.

Double check you have a `prose` class name wrapping the MDX content as noted in https://github.com/remarkjs/react-markdown/issues/505#issuecomment-724911501

If the issue persists.
Use the browser developer tools to see what is being applied.
Right click on the element that looks off and use the style inspector to see where the styles are coming from
* Chrome docs: https://developer.chrome.com/docs/devtools/css/
* Firefox docs: https://firefox-source-docs.mozilla.org/devtools-user/page_inspector/how_to/examine_and_edit_css/index.html

**Reply by ChristianMurphy** [reply]
Also, broad thought/suggestion.
If your goal is to create a blog for easy authoring of posts and low maintaince. (Versus a focus on playing around with Next and Tailwind purely for learning).
Consider moving up a layer of abstract and using https://nextra.site/ with the blog theme.

**Reply by hbhutta** [reply]
Thank you for the suggestion, but I will stick with the current stack because I have found some YouTube tutorials which might I think might help. I will get back to you by tomorrow if there are still any problems. And I *am* actually trying to learn Next.js and Tailwind so I will not use `nextra`.

**Reply by hbhutta** [reply]
> Double check you have a prose class name wrapping the MDX content

I am not doing this at the moment, and I just realized. Right now this is how I'm accessing my MDX files:

```javascript
import BlogCardType from "@/types/BlogCardType";
import Link from "next/link";

/**
 * 
 * @param blog_title
 * @returns Returns the string converted to lowercase with whitespace replaced with '-'
 */
function blogRoute(blog_title: string) {
    return '/blogs/' + blog_title.toLowerCase().replace(/\s/g, '-');
}

export default function BlogCard({ blog_title, published_date, tags} : BlogCardType) { 
    const tagList = tags.map((tag) => <span id='blog-tag' className="blog-tag">{tag}</span>)

    return (
        <div id='blog-card' className="rounded-lg mx-2 my-2">
            <Link href={blogRoute(blog_title)}><span id='blog-title' className="text-4xl font-semibold hover:text-teal-500">{blog_title}</span></Link>
            <span id='blog-title' className="text-4xl font-semibold hover:text-teal-500">{blog_title}</span>
            <div id='blog-info' className="flex justify-between my-3">
                <span id='published-date' className="text-base font-semibold text-gray-400">{published_date}</span>
                <div id='blog-tags'>{tagList}</div>
            </div>
            <hr/>   
        </div>
    )
}
```

Notice that my intent is for the user to click on the span/title of the Blog post and have that blog post load automatically.
```javascript
<Link href={blogRoute(blog_title)}><span id='blog-title' className="text-4xl font-semibold hover:text-teal-500">{blog_title}</span></Link>
```

For the record, this *does* work (i.e. the clicking on the span to load the page), but the rendering of the actual MDX properly (e.g. with all the code blocks, markdown syntax, etc..) -- that's not working still.

Am I missing something here? 

I also used the Developer tools to check if the MDX is being properly mapped to the expected HTML and it is.

**Reply by ChristianMurphy** [reply]
> I also used the Developer tools to check if the MDX is being properly mapped to the expected HTML and it is.

Then MDX is working as expected 🙂 

> And I am actually trying to learn Next.js and Tailwind 

I'm getting the sense from this and your past discussion, that could be more generally phrased as learning to build web applications?
I'd highly recommend a first principles approach, learning JS, CSS, and HTML first:

* https://www.w3schools.com/
* https://javascript.info/
* https://developer.mozilla.org/en-US/docs/Learn

Building on that by learning a framework, like React:

* https://react.dev/learn
* https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Client-side_JavaScript_frameworks/React_getting_started

Before diving into full stack like Remix, Next, Astro etc.

Similarly with Tailwind; start with plain CSS, learn [BEM](https://getbem.com/) and [OOCSS](https://www.smashingmagazine.com/2011/12/an-introduction-to-object-oriented-css-oocss/), before diving into Tailwind or CSS-in-JS frameworks.

Or if you're looking for that organized into courses:

* https://www.freecodecamp.org
* https://www.codecademy.com/
* https://kentcdodds.com/courses
* https://wesbos.com/courses

> Am I missing something here?

1. that an unrelated page shouldn't affect the styles of a blog post itself
2. that next does, or should, provide permalinks to posts, so the regular expression to turn title into a link should not be needed

---

MDX is working as expected, rendering HTML.
Your question is on how to apply Tailwind styles, the Tailwind community can help more in depth, and already has a collection of related questions which may help you https://github.com/tailwindlabs/tailwindcss/discussions?discussions_q=is%3Aopen+mdx
For questions on how to handle routing in Next.js, the Next.js community can help more in depth https://github.com/vercel/next.js/discussions

Reiterating the idea on learning, both Next.js and Tailwind are powerful, but come with a lot of complexity and assume a great deal of knowledge of web development to use effectively. If you are just beginning your web development journey they are not where I would recommend starting from.

**Reply by hbhutta** [reply]
Hi Christian, thank you for the links. You said I need to wrap the MDX content in `prose` (I am going to assume you are not talking about `react-markdown`, and about the Tailwind typography). Then in the code I've shown above, how exactly would I change it to render the MDX properly? I am interested in learning about Next.js and Tailwind, and fixing this specific issue.

**Reply by hbhutta** [reply]
@ChristianMurphy Thank you for your help so far, I really appreciate it! I have one last question, then I will ask somewhere else because I feel that I am taking too much of your time, and I have already spent three days on this one silly problem myself.

I am using the Pages router in NextJS. I am also using TailwindCSS. My nextjs config file contains remark and rehype plugins for syntax highlighting and github flavoured markdown. Here is my project structure again -- nothing wrong with it as far as I can tell: 

![image](https://github.com/mdx-js/mdx/assets/112280419/e09d4364-4c47-460b-af15-45c97ec2b31d)

You can see where I have the .mdx files. 

I have nothing that could be interfering with the mdx rendering in my styles:
![image](https://github.com/mdx-js/mdx/assets/112280419/23ae6102-c129-40cc-b12a-2ea59f5dd92e)

My tailwind config:
```
import type { Config } from 'tailwindcss'

const config: Config = {
  content: [
    './pages/**/*.{js,ts,jsx,tsx,mdx}',
    './components/**/*.{js,ts,jsx,tsx,mdx}',
    './app/**/*.{js,ts,jsx,tsx,mdx}',
  ],
  plugins: [
    require('@tailwindcss/typography')
  ],
}
export default config;
```

The routing is not an issue (putting the mdx files in `../pages/blogs` works). So I think it must be this:
```javascript
import BlogCardType from "@/types/BlogCardType";
import Link from "next/link";

/**
 * 
 * @param blog_title
 * @returns Returns the string converted to lowercase with whitespace replaced with '-'
 */
function blogRoute(blog_title: string) {
    return '/blogs/' + blog_title.toLowerCase().replace(/\s/g, '-');
}

export default function BlogCard({ blog_title, published_date, tags} : BlogCardType) { 
    const tagList = tags.map((tag) => <span id='blog-tag' className="blog-tag">{tag}</span>)

    return (
        <div id='blog-card' className="rounded-lg mx-2 my-2">
            <Link href={blogRoute(blog_title)} className="prose"><span id='blog-title' className="text-4xl font-semibold hover:text-teal-500">{blog_title}</span></Link>
            {/* <span id='blog-title' className="text-4xl font-semibold hover:text-teal-500">{blog_title}</span> */}
            <div id='blog-info' className="flex justify-between my-3">
                <span id='published-date' className="text-base font-semibold text-gray-400">{published_date}</span>
                <div id='blog-tags'>{tagList}</div>
            </div>
            <hr/>   
        </div>
    )
}
```

Specifically this line:
```javascript
<Link href={blogRoute(blog_title)} className="prose"><span id='blog-title' className="text-4xl font-semibold hover:text-teal-500">{blog_title}</span></Link>
```

You can see that I've tried using `prose` but it's still not rendering as it would in, say, the MDX playground on their website, for example. 

Here is a YouTube tutorial styling the MDX with Tailwind, so Tailwind inference is not an issue I don't think.

My question is: Is the specific way I'm trying to access the MDX file as I've shown in the last code snippet above causing the issue potentially? Like, using the Link component on that little string works and brings me to the mdx page, but when I get there it's not rendered properly (I've explained before what I mean by rendering properly).

I really need your help in resolving this problem, please tell me what you think about my code above and about what could be causing this problem. Thanks.

@remcohaszing If you could help me as well I would be really grateful. Thank you!

**Reply by ChristianMurphy** [reply]
> You said I need to wrap the MDX content in prose (I am going to assume you are not talking about react-markdown, and about the Tailwind typography). 

`prose` is a CSS class provided by `tailwind`. https://github.com/remarkjs/react-markdown/issues/505#issuecomment-724911501
You apply it as a CSS class around the content you want to render as `prose`.
You need to apply it around the MDX document with `className="prose"` on the MDX document itself or on a `<div>` wrapping the document.
Most likely on a Next.js layout component https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#layouts

> You can see that I've tried using `prose`

@hbhutta please read some of the materials attached earlier on HTML and CSS.
Putting the style on a link to the document, will not style the document itself, that isn't how links or stylesheets work.

> I really need your help in resolving this problem, please tell me what you think about my code above and about what could be causing this problem. Thanks.

I really need you to read actually read, and not ignore, the past responses.
This has been answered repeatedly.
You need to style the document.

You could use Tailwind's prose, you could apply https://github.com/sindresorhus/github-markdown-css, you could add your own styles.
That is your choice on your site.

The MDX forum is support for MDX.
MDX is rendering the document correctly.

The MDX forum is not your personal tutorial on CSS basics.

There are many wonderful resources to help you on that journey.

* https://www.w3schools.com/
* https://javascript.info/
* https://developer.mozilla.org/en-US/docs/Learn
* https://www.freecodecamp.org
* https://www.codecademy.com/
* https://kentcdodds.com/courses
* https://wesbos.com/courses

**Reply by hbhutta** [reply]
Thanks! You're right. I need to go over some HTML and CSS basics before jumping into Tailwind and Next.

So, just to understand, MDX doesn't style `.mdx` files by default? I was expecting something like this:

![image](https://github.com/mdx-js/mdx/assets/112280419/160d6be0-7997-48ff-b469-f005ea1c3c3b)

**Reply by ChristianMurphy** [reply]
> So, just to understand, MDX doesn't style .mdx files by default?

It does not.
It generates HTML, without any additiinal styles/CSS.

> I was expecting something like this

The MDX website includes a CSS theme, you can try to add the some one to your site if you'd like (it may conflict with tailwind though).
This is the CSS file from the MDX site: https://github.com/mdx-js/mdx/blob/85c5dd058ae2545c82f3b8b64129ad7974d84a90/docs/_asset/index.css

**Reply by hbhutta** [reply]
Ah, that clarifies things. I'm very sorry for taking up so much of your time Christian. I was under the assumptions that all the styling would be done for me! I'll take a look at the MDX website's theme.

Thank you so much for your help -- it looks like I have a lot of learning to do! 

I'll close the discussion.

**Reply by hbhutta** [reply]
Hi Christian, I'm sorry for opening the discussion again. I am looking through  the [mdx.js](https://mdxjs.com/guides/) page, but I cannot find anything on styling other than what I found on the [Nextjs](https://nextjs.org/docs/pages/building-your-application/configuring/mdx) page. I just want to know how I apply the styles on the mdx files. I downloaded this `.css` file but do I have to convert all of that into things like this in my `mdx-components.tsx` file?:

```javascript
import type { MDXComponents } from 'mdx/types'
 
// This file allows you to provide custom React components
// to be used in MDX files. You can import and use any
// React component you want, including inline styles,
// components from other libraries, and more.
 
export function useMDXComponents(components: MDXComponents): MDXComponents {
  return {
    // Allows customizing built-in components, e.g. to add styling.
    h1: ({ children }) => <h1 style={
        {
            fontWeight: '700',
            letterSpacing: '0.0125em'
        }
    }>{children}</h1>,

    pre: ({children}) => <pre style={
        {
            fontSize: 'smaller',
            fontFeatureSettings: 'normal',
            lineHeight: 'calc(1em + (1 / 0.8 * 1ex))'
        }
    }>{children}</pre>,
    ...components,
  }
}
```

Because doing this for each HTML-MDX element mapping would be tedious, right?

**Reply by ChristianMurphy** [reply, marked as answer]
How styles are applied to the page is handled by Next.js
You can read their documentation https://nextjs.org/docs/pages/building-your-application/styling/css-modules#external-stylesheets
and ask their community https://nextjs.org/docs/community

The MDX community is not Next.js support.
